### ISSUE TITLE
Consider changing the concepts of systems and use the Component API 

### ISSUE THREAD

**Issue by ngokevin** [aframevr/aframe#2071]
**Description:**

Systems are currently pretty much just components that auto-initialize on the scene and set a `this.system` reference to components that match the system name. The concept is that they are services or managers for components, but there is not much API or features to it, only semantics.

The issue is that we have to maintain both Systems + Components. This means extra code for us, and extra concepts for developers. And then things like the Inspector and Registry have to take the differences into consideration.

The proposal is to remove `AFRAME.registerSystem` and use `AFRAME.registerComponent`.

## API

The true MVP is to just remove the idea altogether and use components. The developer can specify the API and separation of concerns.

```js
AFRAME.registerComponent('physics-system', {
  addBody: function (el) { // ... }
});

AFRAME.registerComponent('physics-body', {
  init: function () {
    scene.components['physics-system'].addBody(this.el);
  }
});
```

But we could add a property to differentiate a normal component and a "manager" or "system" component. This alone can enforce that the component must be added to `<a-scene>`. And that perhaps auto-initializes once it's registered.

```js
AFRAME.registerComponent('physics-system', {
  system: true
});
```

Taking it further, we can instead specify the components it oversees to give it references and handlers:

```js
AFRAME.registerComponent('physics-system', {
  manages/systemOf: ['physics-body', 'physics-joint'],

  entityAdded: function (el, component) {
    PhysicsWorld.addBody(...);
  },

  entityRemoved: function (el, component) {
    PhysicsWorld.removeBody(...);
  }
});

scene.components['physics-system'].entities
```

Then @dmarcos thought about handling entity updates and marking them as dirty such that system updates in response to a component update are run only once per tick as well.

## Benefits

- Less concepts for A-Frame. There are only Entities and Components.
- Get all of the magic of components (lifecycle, data management). 
- Smaller API surface area. We can delete a bunch of code and documentation and issues.
- The Inspector gets "system" support for free.
- Fixes issues we have currently with systems (no update/play/pause).

**Comment by GavanWilhite** [comment]
This seems pretty reasonable. In the way we use Unity, we simply have "components", and "components that happen to be singletons and are accessed in a static way". Coming from this mindset, it seems more important that there is an easy way to access system instances, rather than some sort of automatic registration or association. 

This could look pretty close to your normal component + flag suggestion where we: 

1) Remove the existing system implementation
2) Adding a  `system: true` flag

Where the flag:

1) Enforces that the component can only go on the scene (effectively enforcing a singleton)
2) Automatically puts the instance of this component on `AFRAME.systems` when it is instantiated (instead of the definition)

Auto-instantiation seems like it might be tricky. I feel like it would be better to let users decide when they have a system on their scene (and this allows for non-optional configuration).

**Comment by wizgrav** [comment]
what @GavanWilhite said plus a "requires: [component, component]" property on components. This will make sure that singletons are created on scene as needed. 

Mind though that systems currently serve not only as singletons but also provide some execution order. At the very least they would have to be separated from common components and run on a different step, as is currently.

**Comment by ngokevin** [comment]
@wizgrav I don't think it's necessary to define any API for execution ordering, the developer could have control over it. If you defined a system that oversees multiple components, you'd have a group of entities that have all of those components, then you can do as you please.

```js
AFRAME.registerComponent('foo-system', {
  systemOf: [['bar', 'baz']],

  tick: function () {
    var barData = this.entities[0].getAttribute('foo');
    var bazData = this.entities[0].getAttribute('baz');

    // 1. Do stuff with barData.
    // 2. Do stuff with bazData.
    // Do whatever.
  }
});
```

**Comment by wizgrav** [comment]
@ngokevin yeah but if foo and baz have ticks() on them it is not guaranteed that the foo-system's tick() will run after the child components, as is the case right now with systems

**Comment by ngokevin** [comment]
That's why you wouldn't put `tick`s on the components. You have the components act as data containers, and delegate the logic/execution to the system.

**Comment by donmccurdy** [comment]
For physics, my components implement a `step()` function, and the system invokes those in custom (two-phase) ordering.

I wouldn't want to reduce systems to unprivileged components without also having real control over `tick()` order. That would potentially break physics badly. Managing `tick()` order seems like a much harder problem. Having systems extend the Component prototype, and support play/pause, would seem great though. Is it possible to merge system/component code while continuing to support a two-pass `tick()`? So there would still be a concept of systems, but fewer codepaths?

I also like the `systemOf` syntax. I could see AFRAME.registerSystem just being an alias of AFRAME.registerComponent that enforces/requires a `systemOf` property.

**Comment by ngokevin** [comment]
We could make systems extend components, but there's a lot of code involved in managing the lifecycle. Right now, the lifecycle is the same so it makes sense to make them components with some extra sauce.

**Comment by donmccurdy** [comment]
The minimum level of extra sauce, IMO, would be that system-components' `tick()` handlers are invoked after all components on scene descendants. Or, alternatively, one of the various tick ordering proposals.

**Comment by donmccurdy** [comment]
What are the open questions for answers/feedback on this currently?

**Comment by ngokevin** [comment]
I'm currently thinking having Systems extend Components but keep the `registerSystem` API. I can imagine Systems gaining more functionality like `entityAdded` and `entityRemoved` handlers.

**Comment by ngokevin** [comment]
I'll close this for now as System updates just landed and there is currently a separate issue for entityAdded/entityRemoved handlers. If refactoring is needed later, we can re-evaluate.
